File: lib/scala/pkg/src/main/scala/org/enso/pkg/validation/NameValidation.scala
package org.enso.pkg.validation

import scala.collection.immutable.ListSet

object NameValidation {

  /** Transforms the given string into a valid package name (i.e. a CamelCased identifier).
    *
    * @param name the original name.
    * @return the transformed name conforming to the specification.
    */
  def normalizeName(name: String): String = {
    val starting =
      if (
        name.isEmpty ||
        name.filter(_ != '_').forall(c => !isAllowedNameCharacter(c))
      ) "Project"
      else if (!name.head.isLetter) "Project_" ++ name
      else name
    val startingWithUppercase = starting.capitalize
    val onlyAlphanumeric      = startingWithUppercase.filter(isAllowedNameCharacter)

    onlyAlphanumeric
  }

  /** Validate the project name.
    *
    * @param name the project name to validate
    * @return either a validation error or a project name if it's valid
    */
  def validateName(name: String): Either[InvalidNameError, String] =
    if (name.isEmpty) {
      Left(InvalidNameError.Empty)
    } else if (!name.head.isLetter || !name.head.isUpper) {
      Left(InvalidNameError.ShouldStartWithCapitalLetter)
    } else if (!name.forall(isAllowedNameCharacter)) {
      val invalidCharacters = name.filterNot(isAllowedNameCharacter)
      Left(
        InvalidNameError.ContainsInvalidCharacters(
          ListSet(invalidCharacters: _*)
        )
      )
    } else {
      Right(name)
    }

  /** Checks if a character is allowed in a project name.
    *
    * @param char the char to validate
    * @return `true` if it's allowed, `false` otherwise
    */
  private def isAllowedNameCharacter(char: Char): Boolean = {
    char.isLetterOrDigit || char == '_'
  }

}

File: app/project-manager-shim/src/projectService/nameValidation.ts
export type InvalidNameError =
  | { type: 'Empty' }
  | { type: 'ShouldStartWithCapitalLetter' }
  | { type: 'ContainsInvalidCharacters'; characters: Set<string> }

/**
 * Checks if a character is allowed in a project name.
 * @param char the char to validate
 * @returns true if it's allowed, false otherwise
 */
function isAllowedNameCharacter(char: string): boolean {
  return /^[a-zA-Z0-9_]$/.test(char)
}

/**
 * Transforms the given string into a valid package name (i.e. a CamelCased identifier).
 * @param name the original name.
 * @returns the transformed name conforming to the specification.
 */
export function normalizedName(name: string): string {
  let starting: string
  if (
    name.length === 0 ||
    name
      .replace(/_/g, '')
      .split('')
      .every((c) => !isAllowedNameCharacter(c))
  ) {
    starting = 'Project'
  } else if (!/^[a-zA-Z]/.test(name)) {
    starting = 'Project_' + name
  } else {
    starting = name
  }
  // Capitalize first letter
  const startingWithUppercase = starting.charAt(0).toUpperCase() + starting.slice(1)
  // Filter to only alphanumeric characters
  const onlyAlphanumeric = startingWithUppercase.split('').filter(isAllowedNameCharacter).join('')

  return onlyAlphanumeric
}

/**
 * Validate the project name.
 * @param name the project name to validate
 * @returns either a validation error or a project name if it's valid
 */
export function validateName(name: string): InvalidNameError | string {
  if (name.length === 0) {
    return { type: 'Empty' }
  }

  if (!/^[A-Z]/.test(name)) {
    return { type: 'ShouldStartWithCapitalLetter' }
  }

  const invalidChars = name.split('').filter((c) => !isAllowedNameCharacter(c))
  if (invalidChars.length > 0) {
    return {
      type: 'ContainsInvalidCharacters',
      characters: new Set(invalidChars),
    }
  }

  return name
}
